### PR TITLE
feat: Add pause/unpause all bridges extrinsics

### PR DESCRIPTION
**Pull request by freddyli7**
- Add pause_all_bridges admin extrinsic
- Add unpause_all_bridges admin extrinsic
- Add pause_all_domains helper
- Refactor unpause_all_domains helper
- Register new extrinsic in the local runtime

### CODE REVIEW

File: bridge/src/lib.rs [hunk 1]
```diff
@@ -849,10 +910,29 @@ pub mod pallet {
 
 		/// unpause all registered domains in the storage
 		fn unpause_all_domains() {
+			// iterate over the registered domain storage first and write all of them into IsPaused
+			// storage
+			for registered_domain_pair in DestDomainIds::<T>::iter() {
+				IsPaused::<T>::insert(registered_domain_pair.0, false);
+			}
```
Comment (tolak) [inline bridge/src/lib.rs:917]:
```suggestion
			IsPaused::<T>::translate(|_, _| Some(false));
```
This way should be more effective

File: bridge/src/lib.rs [hunk 2]
```diff
@@ -849,10 +910,29 @@ pub mod pallet {
...
+			// iterate over all IsPaused storage to set them as false(unpaused)
 			for bridge_pair in IsPaused::<T>::iter() {
```
Comment (tolak) [inline bridge/src/lib.rs:919]: It seems duplicated with above?

**Review by tolak — COMMENTED** [review]
Had some comments, please take a look

File: bridge/src/lib.rs [hunk 3]
```diff
@@ -617,6 +623,61 @@ pub mod pallet {
...
+				Error::<T>::AccessDenied
+			);
+
+			// Pause all bridges
+			Self::pause_all_domains();
+
+			// Emit AllBridgePaused
+			let sender = match ensure_signed(origin) {
+				Ok(sender) => sender,
+				_ => [0u8; 32].into(),
```
[the remark this answers is not in the document]
Comment (freddyli7) [reply bridge/src/lib.rs:646]: good question, I actually copied this from Matija's code in other extrinsic, but now I saw that `has_access` already verified the origin with `ensure_signed`, so I guess this is why he didn't return err, but I think it is always better to have decoupled logic, I will refactor all related part. Sharp eyes bro

File: bridge/src/lib.rs (hunk 1 above)
Comment (freddyli7) [reply bridge/src/lib.rs:917]: I tried with `translate` method but it requires adding `#![recursion_limit = "256"]` attribute to bridge crate, so I did, and it keeps complaining about recursion limit all the way to 131072, then the compile time takes more than 2 hours, so I shut it off. the Error is shown in the below screenshot:
image

I guess it has something to do with the Decode trait that translate is using?

File: bridge/src/lib.rs (hunk 2 above)
Comment (freddyli7) [reply bridge/src/lib.rs:919]: it seems a bit duplicated, but my idea was that iteration over the `registered_domain` storage first and insert all domains into the `IsPaused` storage, then iteration over `IsPaused` to pause/unpause them, this will make sure to sync between `registered_domain` and `IsPaused`. Because when register a new domain A, it only exists in `registered_domain` storage, if we pause all domain at this moment, A will not be pause because it doesn't exist in `IsPaused` storage.

File: bridge/src/benchmarking.rs [hunk 4]
```diff
@@ -244,4 +244,84 @@ mod benchmarks {
 		assert_eq!(Balances::<T, _>::free_balance(caller), 200000000.into());
 		assert_eq!(Balances::<T, _>::free_balance(bridge_account), (amount - 200000000).into());
 	}
+
+	#[benchmark]
+	fn pause_all_bridges() {
```
[the remark this answers is not in the document]
Comment (freddyli7) [reply bridge/src/benchmarking.rs:249]: I see what you meant, to my understanding, this extrinsic is accessing two storages(DestDomainIDs, IsPaused), so I think the benchmark might need to take two length as parameters of each one to perform. And, the size of `IsPaused` is dynamically changing while extrinsic executing because what's in the `DestDomainIDs` will be topped up onto `IsPaused`, so it makes the situation even more complicated. 

Also as I'm looking into it, I found that there is no way to get the size of a `StorageMap` ([parity doc of StorageMap](https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageMap.html#)). Not sure if I'm missing anything here.

in summary, my idea is:
1. IF there is a way to get storageMap size, we add it as the benchmark parameter but only add size of `DestDomainIDs`. This way we will at least simulate the first storage loop logic to make benchmark more accurate, `IsPaused` in the most case will be the subset of the `DestDomainIDs` so we should be fine.
2. IF there is no way to get storageMap size, we hardcode an estimated size in the benchmark, for instance `2`, currently we have 2 domains on mainnet, we can update this number when we adding much more domains later, but for sure this wouldn't be a good methodology.

What you think? @tolak

File: bridge/src/lib.rs (hunk 1 above)
Comment (freddyli7) [reply bridge/src/lib.rs]: @tolak I pushed the fix for this comment, please have another look 🙏

File: bridge/src/benchmarking.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tolak) [reply bridge/src/benchmarking.rs:249]: I also haven't found a way to get the storage size as the argument passes to the benchmark function. Maybe we can implement a dedicated method in pallet but I don't think it's a wise choice because every time ppl call the method it will iterator the map in advance.
As you suggested, let's just hardcode a reasonable number until we find a better solution

File: bridge/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tolak) [reply bridge/src/lib.rs:646]: I see, thanks for clarification

File: bridge/src/lib.rs (hunk 1 above)
Comment (tolak) [reply bridge/src/lib.rs]: Nice job!

File: bridge/src/lib.rs (hunk 2 above)
Comment (tolak) [reply bridge/src/lib.rs]: Get it, thanks for explanation

**Review by tolak — APPROVED** [review]
LGTM
